Azure SQL driver: bad connection

Not certain if this is related in any way with #119 or #105 but here the situation.

I have a major issue where lots of the database call return this error `driver: bad connection`. Here's how I implemented my data access package.

I have a `db` variable that is initiated from the consumer package. So I have my web api which open the database / ping in the main function like this.

```
// consumer package [web api]
func main() {
  dal.OpenDatabase() // <- this initiate and connect / ping to SQL database
}
// dal package
func add(...) {
  stmt, err := db.Prepare(...)
}
```

From what I understand the `sql` package should handle connection pooling, so by using the `Prepare` statement a new connection should be given if for some reason the state of the current one is in a bad state.

Should I instead create a database object for each request in the web project instead of having only one global database connection?

I'm mostly doing something wrong, but at the moment I'm not sure where to start looking to fix this error.

Any help would be highly appreciated,
Thanks

When running tests against an Azure instance, it fails for two tests that tested time outs:

--- FAIL: TestDefaultTimeout (90.70s)
        queries_test.go:311: Exec should fail with timeout, failed with driver: bad connection
--- FAIL: TestShortTimeout (6.65s)
        queries_test.go:332: Exec should fail with timeout, failed with driver: bad connection

I'm getting a high volume of those `bad connection` errors on production and I need to find what's the cause. Does the time out would cause this, so incrementing connection time out could in fact help? Still I would prefer fixing this the right way and not just monkey patch.

Thanks for any pointer.

--- a/mssql.go
+++ b/mssql.go
@@ -31,8 +31,11 @@ func CheckBadConn(err error) error {
 		return driver.ErrBadConn
 	}
 
-	switch err.(type) {
+	switch e := err.(type) {
 	case net.Error:
+		if e.Timeout() {
+			return errors.New("driver: connection timeout")
+		}
 		return driver.ErrBadConn
 	default:
 		return err

